src/gt4py/next/ffront/foast_to_gtir.py

```diff
@@ -0,0 +1,384 @@
+            # operation count if both branches are evaluated.
+            inner_expr_name = self.uid_generator.sequential_id(prefix="__inner_expr")
+            inner_expr_evaluator = im.lambda_(*common_syms)(inner_expr)
+            inner_expr = im.call(inner_expr_name)(*common_symrefs)
+
+            true_branch = self.visit(node.true_branch, inner_expr=inner_expr, **kwargs)
+            false_branch = self.visit(node.false_branch, inner_expr=inner_expr, **kwargs)
+
+            return im.let(inner_expr_name, inner_expr_evaluator)(
+                im.cond(cond, true_branch, false_branch)
+            )
+
+        assert return_kind is foast_introspection.StmtReturnKind.UNCONDITIONAL_RETURN
+
+        # note that we do not duplicate `inner_expr` here since if both branches
+        #  return, `inner_expr` is ignored.
+        true_branch = self.visit(node.true_branch, inner_expr=inner_expr, **kwargs)
+        false_branch = self.visit(node.false_branch, inner_expr=inner_expr, **kwargs)
+
+        return im.cond(cond, true_branch, false_branch)
+
+    def visit_Assign(
+        self, node: foast.Assign, *, inner_expr: Optional[itir.Expr], **kwargs: Any
+    ) -> itir.Expr:
+        return im.let(self.visit(node.target, **kwargs), self.visit(node.value, **kwargs))(
+            inner_expr
+        )
+
+    def visit_Symbol(self, node: foast.Symbol, **kwargs: Any) -> itir.Sym:
+        return im.sym(node.id)
+
+    def visit_Name(self, node: foast.Name, **kwargs: Any) -> itir.SymRef:
+        return im.ref(node.id)
+
+    def visit_Subscript(self, node: foast.Subscript, **kwargs: Any) -> itir.Expr:
+        return im.tuple_get(node.index, self.visit(node.value, **kwargs))
+
+    def visit_TupleExpr(self, node: foast.TupleExpr, **kwargs: Any) -> itir.Expr:
+        return im.make_tuple(*[self.visit(el, **kwargs) for el in node.elts])
+
+    def visit_UnaryOp(self, node: foast.UnaryOp, **kwargs: Any) -> itir.Expr:
+        # TODO(tehrengruber): extend iterator ir to support unary operators
+        dtype = type_info.extract_dtype(node.type)
+        if node.op in [dialect_ast_enums.UnaryOperator.NOT, dialect_ast_enums.UnaryOperator.INVERT]:
+            if dtype.kind != ts.ScalarKind.BOOL:
+                raise NotImplementedError(f"'{node.op}' is only supported on 'bool' arguments.")
+            return self._map("not_", node.operand)
+
+        return self._map(
+            node.op.value,
+            foast.Constant(value="0", type=dtype, location=node.location),
+            node.operand,
+        )
+
+    def visit_BinOp(self, node: foast.BinOp, **kwargs: Any) -> itir.FunCall:
+        return self._map(node.op.value, node.left, node.right)
+
+    def visit_TernaryExpr(self, node: foast.TernaryExpr, **kwargs: Any) -> itir.FunCall:
+        assert (
+            isinstance(node.condition.type, ts.ScalarType)
+            and node.condition.type.kind == ts.ScalarKind.BOOL
+        )
```

Comment:
```suggestion
        assert type_info.is_logical(node.condition.type)
```
I know this function exists, but never how it is called.

Reply:
This is going through the `extract_dtype` and therefore doesn't do the first check. We could introduce for all the cases the `is_XX_field` and `is_XX_scalar`.